File: bundles/org.eclipse.rap.rwt.addons/js/org/eclipse/rap/rwt/addons/camera/Camera.js
/*******************************************************************************
 * Copyright (c) 2025, 2026 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

(function(){
  'use strict';

  rwt.define( "rwt.widgets" );

  var VIDEO_INPUT_DEVICE_ID_KEY = "rwt.widgets.Camera.videoInputDeviceId";

  rwt.widgets.Camera = function( properties ) {
    bindAll( this, [ "init", "layout", "start", "stop", "takePicture", "showNextVideoInputDevice",
                     "isRunning", "onRender", "destroy", "capture" ] );
    this.parent = rap.getObject( properties.parent );
    this.uploadPath = properties.uploadPath;
    this.canvasElement = document.createElement( "canvas" );
    this.videoElement = document.createElement( "video" );
    this.cameraFlipElement = document.createElement( "img" );
    this.cameraCaptureElement = document.createElement( "img" );
    this.parent.append( this.videoElement );
    this.parent.append( this.canvasElement );
    this.parent.append( this.cameraFlipElement );
    this.parent.append( this.cameraCaptureElement );
    this.parent.addListener( "Resize", this.layout );
    this.parent.addListener( "Dispose", this.destroy );
    this.videoInputDeviceId = null;
    this.currentStream = null;
    rap.on( "render", this.onRender );
  };

  rwt.widgets.Camera.prototype = {

    onRender : async function() {
      await this.init();
      this.layout();
      rap.off( "render", this.onRender );
      this.start();
    },

    destroy : function() {
      this.stop();
      this.cameraFlipElement.removeEventListener( "click", this.showNextVideoInputDevice );
      this.cameraCaptureElement.removeEventListener( "click", this.capture );
      this.canvasElement.parentNode.removeChild( this.canvasElement );
      this.videoElement.parentNode.removeChild( this.videoElement );
      this.cameraFlipElement.parentNode.removeChild( this.cameraFlipElement );
      this.cameraCaptureElement.parentNode.removeChild( this.cameraCaptureElement );
    },

    init : async function() {
      this.canvasElement.style.display = "none";
      this.canvasElement.style.position = "absolute";
      this.videoElement.style.position = "absolute";
      this.cameraFlipElement.src = "rwt-resources/camera/camera-flip-32.png";
      this.cameraFlipElement.width = 32;
      this.cameraFlipElement.height = 32;
      this.cameraFlipElement.style.position = "absolute";
      this.cameraFlipElement.style.cursor = "pointer";
      this.cameraFlipElement.style.visibility = "hidden";
      this.cameraFlipElement.addEventListener( "click", this.showNextVideoInputDevice );
      this.cameraCaptureElement.src = "rwt-resources/camera/circle-slice-8-64.png";
      this.cameraCaptureElement.width = 64;
      this.cameraCaptureElement.height = 64;
      this.cameraCaptureElement.style.position = "absolute";
      this.cameraCaptureElement.style.cursor = "pointer";
      this.cameraCaptureElement.style.visibility = "hidden";
      this.cameraCaptureElement.addEventListener( "click", this.capture );
      this.videoInputDeviceId = await initVideoInputDeviceId();
    },

    layout : function() {
      var area = this.parent.getClientArea();
      this.canvasElement.width = area[ 2 ];
      this.canvasElement.height = area[ 3 ];
      this.videoElement.width = area[ 2 ];
      this.videoElement.height = area[ 3 ];
      this.cameraFlipElement.style.top = "5px";
      this.cameraFlipElement.style.left = "20px";
      this.cameraFlipElement.style.visibility = "visible";
      this.cameraCaptureElement.style.top = "40px";
      this.cameraCaptureElement.style.left = "5px";
      this.cameraCaptureElement.style.visibility = "visible";
    },

    start : async function() {
      if( !this.isRunning() ) {
        if( this.videoInputDeviceId != null ) {
          storeVideoInputDeviceId( this.videoInputDeviceId );
          var constraints = {
            video: {
              deviceId: {
                exact: this.videoInputDeviceId
              }
            },
            audio: false
          };
          try {
            this.currentStream = await navigator.mediaDevices.getUserMedia( constraints );
            this.videoElement.autoplay = true;
            this.videoElement.srcObject = this.currentStream;
          } catch( error ) {
            console.error( "Error accessing the camera: ", error );
          };
        }
      }
    },

    stop : function() {
      if( this.isRunning() ) {
        this.currentStream.getTracks().forEach( track => track.stop() );
        try {
          this.videoElement.srcObject = null;
        } catch( error ) {
          this.videoElement.src = "";
        }
        this.videoElement.removeAttribute( "src" );
        this.currentStream = null;
      }
    },

    takePicture : function( properties ) {
      if( this.isRunning() ) {
        var uploadPath = this.uploadPath;
        var video = this.videoElement;
        var canvas = this.canvasElement;
        var context = canvas.getContext( "2d" );
        if( properties.resolution ) {
          canvas.width = properties.resolution[ 0 ];
          canvas.height = properties.resolution[ 1 ];
        } else {
          var aspectRatio = video.videoWidth / video.videoHeight;
          canvas.width = video.width;
          canvas.height = canvas.width / aspectRatio;
        }
        context.drawImage( video, 0, 0, canvas.width, canvas.height );
        canvas.toBlob( ( blob ) => {
          var formData = new FormData();
          formData.append( "image", blob, "camera_picture.jpg" );
          fetch( uploadPath, {
            method: "POST",
            body: formData
          } );
        }, "image/jpeg", properties.compressionQuality );
      }
    },

    capture : function() {
      this.takePicture( { compressionQuality : 1 } );
    },

    showNextVideoInputDevice : async function() {
      var running = this.isRunning();
      if( running ) {
        this.stop();
      }
      this.videoInputDeviceId = await nextVideoInputDeviceId( this.videoInputDeviceId );
      if( running ) {
        this.start();
      }
    },

    isRunning : function() {
      return this.currentStream != null;
    }

  };

  function bindAll( context, methodNames ) {
    for( var i = 0; i < methodNames.length; i++ ) {
      var method = context[ methodNames[ i ] ];
      context[ methodNames[ i ] ] = bind( context, method );
    }
  };

  function bind( context, method ) {
    return function() {
      return method.apply( context, arguments );
    };
  };

  async function initVideoInputDeviceId() {
    // Workaround for Firefox to ask for Camera permissions
    var stream = await navigator.mediaDevices.getUserMedia( { video: true } );
    stream.getTracks().forEach( track => track.stop() );

    var videoInputDeviceId = null;
    var videoInputDevices = await getVideoInputDevices();
    if( videoInputDevices.length > 0 ) {
      videoInputDeviceId = videoInputDevices[ 0 ].deviceId;
    }
    var storedVideoInputDeviceId = rwt.client.Cookie.get( VIDEO_INPUT_DEVICE_ID_KEY );
    for( var i = 0; i < videoInputDevices.length; i++ ) {
      if( videoInputDevices[ i ].deviceId === storedVideoInputDeviceId ) {
        videoInputDeviceId = storedVideoInputDeviceId;
      }
    }
    return videoInputDeviceId;
  };

  async function nextVideoInputDeviceId( currentDeviceId ) {
    var videoInputDeviceId = null;
    var videoInputDevices = await getVideoInputDevices();
    if( videoInputDevices.length > 0 ) {
      videoInputDeviceId = videoInputDevices[ 0 ].deviceId;
    }
    for( var i = 0; i < videoInputDevices.length - 1; i++ ) {
      if( videoInputDevices[ i ].deviceId === currentDeviceId ) {
        videoInputDeviceId = videoInputDevices[ i + 1 ].deviceId;
      }
    }
    return videoInputDeviceId;
  };

  function storeVideoInputDeviceId( currentDeviceId ) {
    rwt.client.Cookie.set( VIDEO_INPUT_DEVICE_ID_KEY, currentDeviceId, 365, "/" );
  };

  async function getVideoInputDevices() {
    var videoInputDevices = [];
    var mediaDevices = await navigator.mediaDevices.enumerateDevices();
    mediaDevices.forEach( mediaDevice => {
      if( mediaDevice.kind === "videoinput" && mediaDevice.deviceId ) {
        videoInputDevices.push( mediaDevice );
      }
    } );
    return videoInputDevices;
  };

  rap.registerTypeHandler( "rwt.widgets.Camera", {

    factory : function( properties ) {
      return new rwt.widgets.Camera( properties );
    },

    properties : [],

    methods : [ "start", "stop", "takePicture" ]

  } );

}());

File: bundles/org.eclipse.rap.rwt.addons/src/org/eclipse/rap/rwt/addons/camera/Camera.java
/*******************************************************************************
 * Copyright (c) 2025 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

package org.eclipse.rap.rwt.addons.camera;

import static org.eclipse.rap.rwt.widgets.WidgetUtil.getId;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.eclipse.rap.fileupload.FileDetails;
import org.eclipse.rap.fileupload.FileUploadEvent;
import org.eclipse.rap.fileupload.FileUploadHandler;
import org.eclipse.rap.fileupload.FileUploadListener;
import org.eclipse.rap.fileupload.FileUploadReceiver;
import org.eclipse.rap.json.JsonArray;
import org.eclipse.rap.json.JsonObject;
import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.client.service.ClientFileLoader;
import org.eclipse.rap.rwt.remote.Connection;
import org.eclipse.rap.rwt.remote.RemoteObject;
import org.eclipse.rap.rwt.service.ResourceManager;
import org.eclipse.rap.rwt.service.ServerPushSession;
import org.eclipse.rap.rwt.service.UISessionEvent;
import org.eclipse.rap.rwt.service.UISessionListener;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Image;
import org.eclipse.swt.graphics.Point;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Display;

/**
 * Camera control that shows a video stream from the client camera. *
 *
 * @since 4.4
 */
@SuppressWarnings("serial")
public class Camera extends Composite {
  
  private static final Pattern SERVICE_URL_PATTERN = Pattern.compile(".*/([^/.]*\\?.*)"); //$NON-NLS-1$
  
  public static final String PROPERTY_PARENT = "parent";
  public static final String PROPERTY_UPLOAD_PATH = "uploadPath";
  public static final String PROPERTY_RESOLUTION = "resolution";
  public static final String PROPERTY_COMPRESSON_QUALITY = "compressionQuality";
  public static final String METHOD_TAKE_PICTURE = "takePicture";
  
  private static final String JS_PATH = "org/eclipse/rap/rwt/addons/camera/";
  private static final String REGISTER_PATH = "camera/";

  private static final String[] JS_FILES = { "Camera.js" };
  private static final String REMOTE_TYPE = "rwt.widgets.Camera";
  
  private final RemoteObject remoteObject;
  private final List<CameraListener> cameraListeners;
  private final ServerPushSession serverPush;

  public Camera( Composite parent ) {
    super( parent, SWT.NONE );
    registerResources();
    loadJavaScript();
    Connection connection = RWT.getUISession().getConnection();
    remoteObject = connection.createRemoteObject( REMOTE_TYPE );
    remoteObject.set( PROPERTY_PARENT, getId( this ) );
    cameraListeners = new ArrayList<CameraListener>();
    serverPush = new ServerPushSession();
    String uploadPath = registerFileUploadServiceHandler();
    remoteObject.set( PROPERTY_UPLOAD_PATH, stripContextPath( uploadPath ) );
  }
  
  /**
   * <p>
   * Instructs the client to take a picture from the camera. The added {@link CameraListener}s will be called when the
   * user has taken a picture..
   * </p>
   *
   * @see CameraListener
   */
  public void takePicture( CameraOptions options ) {
    checkWidget();
    if( options == null ) {
      SWT.error( SWT.ERROR_NULL_ARGUMENT );
    }
    remoteObject.call( METHOD_TAKE_PICTURE, createProperties( options ) );
    serverPush.start();
  }
  
  /**
   * <p>
   * Adds a {@link CameraListener} to get notified about image events.
   * </p>
   */
  public void addCameraListener( CameraListener listener ) {
    checkWidget();
    if( listener == null ) {
      SWT.error( SWT.ERROR_NULL_ARGUMENT );
    }
    cameraListeners.add( listener );
  }

  /**
   * <p>
   * Removes a {@link CameraListener}.
   * </p>
   */
  public void removeCameraListener( CameraListener listener ) {
    checkWidget();
    if( listener == null ) {
      SWT.error( SWT.ERROR_NULL_ARGUMENT );
    }
    cameraListeners.remove( listener );
  }
  
  @Override
  public void dispose() {
    if( !isDisposed() ) {
      remoteObject.destroy();
    }
    super.dispose();
  }
  
  RemoteObject getRemoteObject() {
    return remoteObject;
  }
  
  private static void registerResources() {
    ResourceManager resourceManager = RWT.getResourceManager();
    for( String fileName : JS_FILES ) {
      registerFileIfNeeded( resourceManager, fileName );
    }
  }

  private static void registerFileIfNeeded( ResourceManager resourceManager, String fileName ) {
    boolean isRegistered = resourceManager.isRegistered( REGISTER_PATH + fileName );
    if( !isRegistered ) {
      try {
        register( resourceManager, fileName );
      } catch( IOException ioe ) {
        throw new IllegalArgumentException( "Failed to load resources", ioe );
      }
    }
  }
  
  private static void loadJavaScript() {
    ClientFileLoader loader = RWT.getClient().getService( ClientFileLoader.class );
    ResourceManager resourceManager = RWT.getResourceManager();
    loader.requireJs( resourceManager.getLocation( REGISTER_PATH + "Camera.js" ) );
  }
  
  private static void register( ResourceManager resourceManager, String fileName ) throws IOException {
    ClassLoader classLoader = Camera.class.getClassLoader();
    InputStream inputStream = classLoader.getResourceAsStream( JS_PATH + fileName );
    try {
      resourceManager.register( REGISTER_PATH + fileName, inputStream );
    } finally {
      inputStream.close();
    }
  }
  
  private static JsonObject createProperties( CameraOptions options ) {
    JsonObject properties = new JsonObject();
    addResolution( properties, options );
    addCompressionQuality( properties, options );
    return properties;
  }
  
  private static void addResolution( JsonObject properties, CameraOptions options ) {
    Point resolution = options.getResolution();
    if( resolution != null ) {
      JsonArray jsonArray = new JsonArray();
      jsonArray.add( resolution.x );
      jsonArray.add( resolution.y );
      properties.add( PROPERTY_RESOLUTION, jsonArray );
    }
  }

  private static void addCompressionQuality( JsonObject properties, CameraOptions options ) {
    properties.add( PROPERTY_COMPRESSON_QUALITY, options.getCompressionQuality() );
  }
  
  void handleUploadFailed( Display display, final ImageUploadReceiver receiver ) {
    if( display != null && !display.isDisposed() ) {
      display.asyncExec( new Runnable() {
        @Override
        public void run() {
          notifyListenersWithoutPicture();
          receiver.reset();
          serverPush.stop();
        }
      } );
    }
  }

  void handleUploadFinished( Display display, final ImageUploadReceiver receiver ) {
    if( display != null && !display.isDisposed() ) {
      display.asyncExec( new Runnable() {
        @Override
        public void run() {
          notifyListenersWithImage( receiver.getImage() );
          receiver.reset();
          serverPush.stop();
        }
      } );
    }
  }

  private void notifyListenersWithImage( Image image ) {
    List<CameraListener> listeners = new ArrayList<CameraListener>( cameraListeners );
    for( CameraListener listener : listeners ) {
      listener.receivedPicture( image );
    }
  }

  private void notifyListenersWithoutPicture() {
    List<CameraListener> listeners = new ArrayList<CameraListener>( cameraListeners );
    for( CameraListener listener : listeners ) {
      listener.receivedPicture( null );
    }
  }

  private String registerFileUploadServiceHandler() {
    Display display = getDisplay();
    ImageUploadReceiver receiver = new ImageUploadReceiver( display );
    FileUploadHandler uploadHandler = new FileUploadHandler( receiver );
    uploadHandler.addUploadListener( new FileUploadListener() {
      @Override
      public void uploadProgress( FileUploadEvent event ) {
      }
      @Override
      public void uploadFailed( FileUploadEvent event ) {
        handleUploadFailed( display, receiver );
      }
      @Override
      public void uploadFinished( FileUploadEvent event ) {
        handleUploadFinished( display, receiver );
      }
    } );
    RWT.getUISession().addUISessionListener( new UISessionListener() {
      @Override
      public void beforeDestroy( UISessionEvent event ) {
        uploadHandler.dispose();
      }
    } );
    return uploadHandler.getUploadUrl();
  }

  static String stripContextPath( String serviceHandlerUrl ) {
    Matcher matcher = SERVICE_URL_PATTERN.matcher( serviceHandlerUrl );
    if( matcher.matches() ) {
      return matcher.group( 1 );
    }
    return serviceHandlerUrl;
  }
  
  public class ImageUploadReceiver extends FileUploadReceiver {

    private final Display display;
    private Image image;

    public ImageUploadReceiver( Display display ) {
      this.display = display;
    }

    @Override
    public void receive( InputStream stream, FileDetails details ) throws IOException {
      image = new Image( display, stream );
    }

    public Image getImage() {
      return image;
    }

    public void reset() {
      image = null;
    }

  }
  
}
